### ISSUE TITLE
tracing: Jaeger not receiving any trace spans with latest Istio

### ISSUE THREAD

**Issue by yangminzhu** [istio/istio#32026]
(**NOTE**: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>)

**Bug description**
I installed the Jaeger (https://istio.io/latest/docs/ops/integrations/jaeger/) with the latest Istio master branch on top of https://github.com/istio/istio/pull/32011 and set the sampling to 100. I started to send requests continuously but could never see any tracing spans in the dashboard.

I then deleted Jaeger, installed the Zipkin (https://istio.io/latest/docs/ops/integrations/zipkin/). I can immediately find the tracing spans showing in Zipkin dashboard. This makes me wondering something seems wrong with the Jaeger integration in Istio.

[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[x] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

**Steps to reproduce the bug**

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

**How was Istio installed?**

**Environment where the bug was observed (cloud vendor, OS, etc)**

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

**Comment by bianpengyuan** [comment]
Sigh.. it looks like https://github.com/istio/istio/issues/22968 is coming back again：

```
4.POST /api/v2/spans HTTP/1.1
host: zipkin
content-type: application/json
x-envoy-internal: true
x-forwarded-for: 10.16.4.17
x-envoy-expected-rq-timeout-ms: 5000
transfer-encoding: chunked

955
[{"kind":"CLIENT","annotations":[{"timestamp":"1618250764089199","value":"cr"}],"tags":{"node_id":"sidecar~10.16.4.17~reviews-v3-58d5f68fc-jwxc9.default~default.svc.cluster.local","request_size":"0","peer.address":"10.16.4.17","http.method":"GET","http.protocol":"HTTP/1.1","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36","istio.canonical_service":"reviews","http.status_code":"200","istio.mesh_id":"cluster.local","response_flags":"-","http.url":"http://ratings:9080/ratings/0","upstream_cluster":"outbound|9080||ratings.default.svc.cluster.local","component":"proxy","istio.namespace":"default","upstream_cluster.name":"outbound|9080||ratings.default.svc.cluster.local","istio.canonical_revision":"v3","response_size":"48","downstream_cluster":"-","guid:x-request-id":"5c2e3166-fc00-95e3-bd05-c6452bd82145","zone":"us-central1-c"},"traceId":"4b72429b22361fa3cf6778d10d68a0d4","id":"547eda649805817d","timestamp":1618250764084480,"duration":4567,"localEndpoint":{"port":0,"serviceName":"reviews.default","ipv4":"10.16.4.17"},"name":"ratings.default.svc.cluster.local:9080/*","parentId":"45b1e963d7114474"},{"kind":"CLIENT","localEndpoint":{"serviceName":"reviews.default","port":0,"ipv4":"10.16.4.17"},"timestamp":1618250765173948,"traceId":"a22e43681be36320e3dc07c8c5fa5e1c","name":"ratings.default.svc.cluster.local:9080/*","tags":{"node_id":"sidecar~10.16.4.17~reviews-v3-58d5f68fc-jwxc9.default~default.svc.cluster.local","http.protocol":"HTTP/1.1","istio.mesh_id":"cluster.local","istio.canonical_revision":"v3","component":"proxy","guid:x-request-id":"47f83075-0ea1-9b17-bc70-2ccf6ccd3f5a","response_size":"48","upstream_cluster.name":"outbound|9080||ratings.default.svc.cluster.local","peer.address":"10.16.4.17","http.url":"http://ratings:9080/ratings/0","istio.canonical_service":"reviews","response_flags":"-","downstream_cluster":"-","request_size":"0","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36","zone":"us-central1-c","istio.namespace":"default","http.method":"GET","http.status_code":"200","upstream_cluster":"outbound|9080||ratings.default.svc.cluster.local"},"annotations":[{"value":"cr","timestamp":"1618250765176839"}],"parentId":"fa23c73086c3679c","id":"dfecce5fbccd7472","duration":2762}]
0

�t`)
ttb�P�Edz
         @@�7
!
$ê���1`b���K�
�
G���=�HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Vary: Origin
X-Content-Type-Options: nosniff
Date: Mon, 12 Apr 2021 18:06:08 GMT
Content-Length: 130

Unable to process request body: json: cannot unmarshal string into Go struct field Annotation.annotations.timestamp of type int64
```

**Comment by bianpengyuan** [comment]
cc @douglas-reid

**Comment by douglas-reid** [comment]
@bianpengyuan how did this come back? didn't we add tests somewhere (including envoy code) for timestamp formatting?

**Comment by bianpengyuan** [comment]
It is a different field this time. That issue is about timestamp in span (which is correctly filled as int64 in on the dump). This time it is the timestamp in annotation.

**Comment by bianpengyuan** [comment]
cc @objectiser in case you know any related changes.

**Comment by bianpengyuan** [comment]
Confirmed that this is not happening at 1.9, so this is a regression.

I marked this as a release blocker. cc @Monkeyanator

**Comment by douglas-reid** [comment]
@bianpengyuan do we need to (possibly) switch default zipkin formats as workaround ?

**Comment by bianpengyuan** [comment]
This is an accepted trace report from 1.9 proxy, which doest no have any annotations:

```
[{"duration":50409,"kind":"CLIENT","traceId":"c7c0d4ff38c181a8f7f8b640927b0406","tags":{"upstream_cluster":"outbound|9080||productpage.default.svc.cluster.local","istio.namespace":"istio-system","node_id":"router~10.16.1.29~istio-ingressgateway-6dbbbccdc8-ns7jp.istio-system~istio-system.svc.cluster.local","http.url":"http://35.226.70.195/productpage","peer.address":"10.44.40.28","request_size":"0","http.method":"GET","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36","http.status_code":"200","istio.mesh_id":"cluster.local","guid:x-request-id":"89ce88e0-be78-930d-b622-53abd477f063","response_size":"5179","response_flags":"-","downstream_cluster":"-","http.protocol":"HTTP/1.1","istio.canonical_service":"istio-ingressgateway","component":"proxy","zone":"us-central1-c","istio.canonical_revision":"latest"},"timestamp":1618258079842878,"id":"f7f8b640927b0406","name":"productpage.default.svc.cluster.local:9080/productpage","localEndpoint":{"ipv4":"10.16.1.29","port":0,"serviceName":"istio-ingressgateway"}},{"name":"productpage.default.svc.cluster.local:9080/productpage","tags":{"peer.address":"10.44.40.28","http.status_code":"200","request_size":"0","downstream_cluster":"-","response_flags":"-","response_size":"5183","http.url":"http://35.226.70.195/productpage","node_id":"router~10.16.1.29~istio-ingressgateway-6dbbbccdc8-ns7jp.istio-system~istio-system.svc.cluster.local","istio.namespace":"istio-system","guid:x-request-id":"78d34746-876c-9105-9fff-d6cf7c2b2620","upstream_cluster":"outbound|9080||productpage.default.svc.cluster.local","component":"proxy","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36","istio.canonical_revision":"latest","zone":"us-central1-c","istio.canonical_service":"istio-ingressgateway","istio.mesh_id":"cluster.local","http.protocol":"HTTP/1.1","http.method":"GET"},"kind":"CLIENT","traceId":"891502324e82df9a073bc69187c0bc73","localEndpoint":{"ipv4":"10.16.1.29","serviceName":"istio-ingressgateway","port":0},"duration":39241,"timestamp":1618258080521282,"id":"073bc69187c0bc73"},{"duration":38167,"name":"productpage.default.svc.cluster.local:9080/productpage","localEndpoint":{"ipv4":"10.16.1.29","serviceName":"istio-ingressgateway","port":0},"id":"352b88b0275083dd","timestamp":1618258081022101,"kind":"CLIENT","tags":{"upstream_cluster":"outbound|9080||productpage.default.svc.cluster.local","zone":"us-central1-c","peer.address":"10.44.40.28","downstream_cluster":"-","guid:x-request-id":"1f11bafc-eb48-9846-952e-2f3ca1b40898","response_size":"5179","node_id":"router~10.16.1.29~istio-ingressgateway-6dbbbccdc8-ns7jp.istio-system~istio-system.svc.cluster.local","component":"proxy","request_size":"0","http.url":"http://35.226.70.195/productpage","response_flags":"-","http.protocol":"HTTP/1.1","istio.namespace":"istio-system","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36","istio.canonical_service":"istio-ingressgateway","istio.canonical_revision":"latest","http.status_code":"200","istio.mesh_id":"cluster.local","http.method":"GET"},"traceId":"923045ed68ca7908352b88b0275083dd"},{"kind":"CLIENT","localEndpoint":{"serviceName":"istio-ingressgateway","port":0,"ipv4":"10.16.1.29"},"id":"b122a53ac2019e60","timestamp":1618258081438232,"duration":33448,"tags":{"http.protocol":"HTTP/1.1","response_flags":"-","guid:x-request-id":"597c91f9-bc80-993e-8ff9-7c45714057c4","http.method":"GET","http.url":"http://35.226.70.195/productpage","node_id":"router~10.16.1.29~istio-ingressgateway-6dbbbccdc8-ns7jp.istio-system~istio-system.svc.cluster.local","peer.address":"10.44.40.28","istio.canonical_revision":"latest","upstream_cluster":"outbound|9080||productpage.default.svc.cluster.local","zone":"us-central1-c","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36","http.status_code":"200","istio.namespace":"istio-system","istio.mesh_id":"cluster.local","response_size":"5183","downstream_cluster":"-","istio.canonical_service":"istio-ingressgateway","component":"proxy","request_size":"0"},"traceId":"04877b346b89e3b8b122a53ac2019e60","name":"productpage.default.svc.cluster.local:9080/productpage"}]

```

**Comment by bianpengyuan** [comment]
> do we need to (possibly) switch default zipkin formats as workaround ?

@douglas-reid sorry what is the default zipkin format? I don't think I know that concept.

**Comment by douglas-reid** [comment]
> > do we need to (possibly) switch default zipkin formats as workaround ?
> 
> @douglas-reid sorry what is the default zipkin format? I don't think I know that concept.

IIRC, we are not setting format in the config, potentially resulting in HTTP_JSON_V1 being used. I'm wondering if HTTP_JSON_V2 could be used here (iirc, that has the patched code) to avoid an Envoy PR for a deprecated option.

**Comment by bianpengyuan** [comment]
I think we are using HTTP_JSON v2? This is the tracing configuration from my cluster:
```json
"tracing": {
     "http": {
      "name": "envoy.tracers.zipkin",
      "typed_config": {
       "@type": "type.googleapis.com/envoy.config.trace.v3.ZipkinConfig",
       "collector_cluster": "zipkin",
       "collector_endpoint": "/api/v2/spans",
       "trace_id_128bit": true,
       "shared_span_context": false,
       "collector_endpoint_version": "HTTP_JSON"
      }
     }
    }
```

I think the problem is somehow envoy starts to add annotations to the trace report, with timestamp as a string value.
1.9 trace span:

```json
{
    "duration": 50409,
    "kind": "CLIENT",
    "traceId": "c7c0d4ff38c181a8f7f8b640927b0406",
    "tags": {
        "upstream_cluster": "outbound|9080||productpage.default.svc.cluster.local",
        "istio.namespace": "istio-system",
        "node_id": "router~10.16.1.29~istio-ingressgateway-6dbbbccdc8-ns7jp.istio-system~istio-system.svc.cluster.local",
        "http.url": "http://35.226.70.195/productpage",
        "peer.address": "10.44.40.28",
        "request_size": "0",
        "http.method": "GET",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36",
        "http.status_code": "200",
        "istio.mesh_id": "cluster.local",
        "guid:x-request-id": "89ce88e0-be78-930d-b622-53abd477f063",
        "response_size": "5179",
        "response_flags": "-",
        "downstream_cluster": "-",
        "http.protocol": "HTTP/1.1",
        "istio.canonical_service": "istio-ingressgateway",
        "component": "proxy",
        "zone": "us-central1-c",
        "istio.canonical_revision": "latest"
    },
    "timestamp": 1618258079842878,
    "id": "f7f8b640927b0406",
    "name": "productpage.default.svc.cluster.local:9080/productpage",
    "localEndpoint": {
        "ipv4": "10.16.1.29",
        "port": 0,
        "serviceName": "istio-ingressgateway"
    }
}
```

vs head trace span:
```json
{
    "parentId": "7ea5f8b6ac2f8db0",
    "name": "productpage.default.svc.cluster.local:9080/static*",
    "traceId": "11458bcbbc4f68957ea5f8b6ac2f8db0",
    "tags": {
        "istio.canonical_service": "productpage",
        "istio.namespace": "default",
        "node_id": "sidecar~10.16.4.18~productpage-v1-5c5d6b66f5-ntfgj.default~default.svc.cluster.local",
        "http.status_code": "200",
        "http.url": "http://35.226.70.195/static/bootstrap/fonts/glyphicons-halflings-regular.woff2",
        "zone": "us-central1-c",
        "upstream_cluster": "inbound|9080||",
        "downstream_cluster": "-",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36",
        "component": "proxy",
        "response_size": "18028",
        "peer.address": "10.16.1.29",
        "http.protocol": "HTTP/1.1",
        "istio.mesh_id": "cluster.local",
        "http.method": "GET",
        "response_flags": "-",
        "request_size": "0",
        "istio.canonical_revision": "v1",
        "upstream_cluster.name": "inbound|9080||",
        "guid:x-request-id": "b7ab9fb4-0c76-9871-892c-60892aed458d"
    },
    "duration": 4260,
    "kind": "SERVER",
    "id": "ce5129e396120bb4",
    "timestamp": 1618258295649178,
    "annotations": [
        {
            "timestamp": "1618258295653722",
            "value": "ss"
        }
    ],
    "localEndpoint": {
        "port": 0,
        "ipv4": "10.16.4.18",
        "serviceName": "productpage.default"
    }
}
```

**Comment by douglas-reid** [comment]
@bianpengyuan ok. sounds reasonable. it looks like there is an endpoint set, so I would expect the following code to work:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/tracers/zipkin/span_buffer.cc#L140-L153

**Comment by douglas-reid** [comment]
oh, maybe the replacement for `timestamp` is overloaded. oh boy. what a pain.

**Comment by bianpengyuan** [comment]
Likely due to https://github.com/envoyproxy/envoy/pull/15492/files.

**Comment by bianpengyuan** [comment]
Tracking bug at Envoy https://github.com/envoyproxy/envoy/issues/15960

**Comment by douglas-reid** [comment]
Upstream fix was: https://github.com/envoyproxy/envoy/pull/16073.  @lizan can you coordinate getting this into the 1.10 release?

**Comment by lizan** [comment]
@dmitri-d lets backport https://github.com/envoyproxy/envoy/pull/16073 to 1.18.x and then point istio/proxy there.

**Comment by bianpengyuan** [comment]
This has been fixed at Istio master and release-1.10.

**Comment by HarshadRanganathan** [comment]
Is this released in Istio 1.19.x?

**Comment by bianpengyuan** [comment]
> Is the fix released in Istio 1.19.x?

You mean Istio 1.9.x? This does not happen on 1.9, but only happens at master, which has been fixed at the to-be released 1.10 branch.

**Comment by HarshadRanganathan** [comment]
I'm using Istio 1.9.2 and observing the same behaviour. No trace spans in Jaegar but on installing Zipkin I can see the traces.

**Comment by HarshadRanganathan** [comment]
Jaeger tracing works when I use port 14268 instead of 9411. Should the docs be corrected?

https://istio.io/latest/docs/ops/integrations/jaeger/#installation

_Once Jaeger is installed, you will need to point Istio proxies to send traces to the deployment. This can be configured with --set values.global.tracer.zipkin.address=<jaeger-collector-address>:9411 at installation time. See the ProxyConfig.Tracing for advanced configuration such as TLS settings._

**Comment by douglas-reid** [comment]
@HarshadRanganathan there is no jaeger-specific serialization supported, so it should be using whatever endpoint is serving the zipkin API. it should work with `:9411` (with the newest envoy patches), using the sample deployment. Are you using a custom deployment?

**Comment by HarshadRanganathan** [comment]
> @HarshadRanganathan there is no jaeger-specific serialization supported, so it should be using whatever endpoint is serving the zipkin API. it should work with `:9411` (with the newest envoy patches), using the sample deployment. Are you using a custom deployment?

Thanks, my bad, it works with `:9411`. Two things resulted in the confusion -

[1] I saw in this comment https://github.com/istio/istio/issues/32596#issuecomment-831429151 that the port should be `14268` and not `9411`

[2] Also, I'm using helm chart for updating istio. So, it seems that envoy doesn't do hot reload of the new config (need to check if I need to add a setting or operator is needed for it) so it stayed with the old config.

And couple of other things that I was doing, led me to believe it was due to port change.

I have checked the envoy proxy logs and can confirm that with port :9411 traces are being published to Jaegar.

Apologies again, for the confusion on my part.
